src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,205 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using NuGet.Common;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol.Model;
+
+namespace NuGet.Protocol.Resources
+{
+    /// <summary>Implementation of <see cref="IVulnerabilityInfoResource"/> for NuGet V3 HTTP feeds.</summary>
+    /// <remarks>Not intended to be created directly. Use <see cref="SourceRepository.GetResourceAsync{T}(CancellationToken)"/>
+    /// with <see cref="IVulnerabilityInfoResource"/> for T, and typecast to this class.
+    /// <para>Implements the <a href="https://learn.microsoft.com/nuget/api/vulnerability-info-resource">VulnerabilityInfo server API resource</a>.</para></remarks>
+    public sealed class VulnerabilityInfoResourceV3 : IVulnerabilityInfoResource
+    {
+        private readonly SourceRepository _sourceRepository;
+
+        internal VulnerabilityInfoResourceV3(SourceRepository sourceRepository)
+        {
+            _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
+        }
+
+        /// <summary>Get the vulnerability pages the server contains.</summary>
+        /// <param name="cacheContext">The cache settings to use when making HTTP requests.</param>
+        /// <param name="log">The logger for any messages.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>The list of vulnerability data pages the server claims to have.</returns>
+        /// <exception cref="FatalProtocolException">When various HTTP or deserialization exceptions occur.</exception>
+        public async Task<IReadOnlyList<V3VulnerabilityIndexEntry>> GetVulnerabilityFilesAsync(SourceCacheContext cacheContext, ILogger log, CancellationToken cancellationToken)
+        {
+            Uri vulnerabilityIndexUrl = await GetIndexUrlAsync(cancellationToken);
+            HttpSourceResource httpSourceResource = await _sourceRepository.GetResourceAsync<HttpSourceResource>(cancellationToken);
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(vulnerabilityIndexUrl.OriginalString, "vuln_index", httpSourceCacheContext);
+            IReadOnlyList<V3VulnerabilityIndexEntry>? vulnFiles;
+            try
+            {
+                vulnFiles = await httpSourceResource.HttpSource.GetAsync(request,
+                    result =>
+                {
+                    using (var textReader = new StreamReader(result.Stream))
+                    using (var jsonReader = new JsonTextReader(textReader))
+                    {
+                        var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IReadOnlyList<V3VulnerabilityIndexEntry>>(jsonReader);
```

Comment:
Do you know if this call will have any effect on Visual Studio?  Will we end up scheduling this on a background thread and not have to worry about it?  Just curious the implications on VS.

Reply:
Are you referring to the fact that it's not async, or something else?

Ideally NuGet.Protocol should never run on the UI thread. But NuGet.Protocol is a "generic" library that runs in, and ships with, CLI msbuild and the .NET SDK. Therefore, we can't use the vs-threading library and `ThreadHelper.ThrowWhenOnUIThread()`. If you have an idea of some other way we can easily detect it (and future regressions), that'll be great (I'm thinking something around ETW events and a post-test analyzer, similar to VS's DDRITs/RPS tests, but this involves a lot of infrastructure NuGet doesn't currently have). But ASP.NET (non-core) has thread-affinity, so we probably don't want to throw any time the task's synchronization context is non-null, because that'll break anyone who might be using NuGet.Protocol in an ASP.NET app (well, it'll force them to be more explicit about switching between the thread pool, and the appropriate synchronization context).

Anyway, Newtonsoft.Json doesn't have an async deserialize method. The only async code I'm aware of is JsonTextReader has a `NextAsync`, but that means we'd have to convert everything to custom converters (and not classes implementing `JsonConverter`). It'd be easier to switch to System.Text.Json instead, although since that only supports UTF-8, there's theoretically a risk of breaking any server that serves any other text encoding. All I can do now is throw my hands in the air 🤷‍♂️

If you're not referring to async, then you need to explain what specific concerns about VS you have.  But it's not fundamentally deserializing JSON differently to how the rest of NuGet.Protocol is already deserializing JSON. Well, it is, most of NuGet.Protocl is much more awful, first converting to JObject, and then deserializing, but that's a pointless performance waste.

Reply:
I was just thinking that most other calls aren't potentially huge and just wanted to be sure that this moderately intense operation won't cause UI blocking in VS.  I'm not familiar enough with JTF to know, just wanted to get your thoughts.  Sounds like we're good.